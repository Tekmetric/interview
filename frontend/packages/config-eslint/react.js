module.exports = {
  extends: [
    '@vercel/style-guide/eslint/browser',
    '@vercel/style-guide/eslint/typescript',
    '@vercel/style-guide/eslint/react'
  ].map(require.resolve),
  parserOptions: {
    project: '**/tsconfig.json',
    sourceType: 'module'
  },
  globals: {
    JSX: true
  },
  settings: {
    'import/resolver': {
      typescript: {
        project: '**/tsconfig.json',
        sourceType: 'module'
      },
      node: {
        extensions: ['.mjs', '.js', '.jsx', '.ts', '.tsx']
      }
    }
  },
  ignorePatterns: ['node_modules/', 'dist/', '.eslintrc.js', '**/*.css'],
  // add rules configurations here
  rules: {
    'import/no-default-export': 'off',
    'import/no-relative-packages': 'off',
    'react/button-has-type': 'off',
    'no-restricted-imports': [
      'error',
      {
        patterns: [
          {
            group: ['packages/*'],
            message: "Please use '@tekmetric/*' imports instead"
          }
        ]
      }
    ],
    'import/order': [
      'error',
      {
        'newlines-between': 'always',
        groups: [
          ['builtin', 'external'],
          'unknown',
          ['parent', 'sibling', 'index']
        ],
        pathGroups: [
          {
            pattern: '~/**',
            group: 'unknown',
            position: 'after'
          }
        ]
      }
    ],
    'react/function-component-definition': [
      'error',
      {
        namedComponents: 'arrow-function',
        unnamedComponents: 'arrow-function'
      }
    ]
  },
  overrides: [
    {
      files: ['*.config.js'],
      env: {
        node: true
      }
    },
    {
      files: ['*.ts', '*.tsx'],
      rules: {
        'no-unused-vars': 'off',
        '@typescript-eslint/no-invalid-void-type': 'off',
        'react/jsx-no-leaked-render': 'off',
        '@typescript-eslint/no-unused-vars': [
          'error',
          {
            vars: 'all',
            args: 'after-used',
            ignoreRestSiblings: true,
            argsIgnorePattern: '^_',
            varsIgnorePattern: '^_'
          }
        ],
        '@typescript-eslint/no-misused-promises': [
          'error',
          {
            checksVoidReturn: false
          }
        ]
      }
    },
    {
      files: ['*.test.ts', '*.test.tsx'],
      env: {
        jest: true
      },
      rules: {
        '@typescript-eslint/no-unsafe-call': 'off',
        '@typescript-eslint/no-unsafe-member-access': 'off'
      }
    }
  ]
}
