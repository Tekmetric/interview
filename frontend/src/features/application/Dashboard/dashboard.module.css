.dashboard {
  --dashboard-button-width: 50px;

  background-color: var(--background-color);
  display: flex;
  height: 100vh;
  width: 100%;
  overflow: hidden;
}

.toggle {
  height: auto;
  width: var(--dashboard-button-width);
}

.sidebar {
  display: flex;
  flex-direction: row-reverse;
  height: 100%;
  position: absolute;
  transform: translateX(calc(-100% + var(--dashboard-button-width)));
  transition: 0.25s linear transform;
  width: 100%;
  z-index: 1;
}

.sidebar.active {
  transform: translateX(0%);
}

.logo {
  padding: var(--space-md);
  display: inline-block;
  color: white;
  text-decoration: none;
}

.logo span {
  display: inline-block;
  font-size: var(--font-size-xl);
  line-height: var(--line-height-xl);
}
.logo span:nth-child(2) {
  font-size: var(--font-size-lg);
  opacity: 0.5;
}

.sidebar__profile {
  padding: var(--space-md);
  text-align: left;
  display: flex;
  flex-direction: column;
  grid-gap: 1rem;
}

.sidebar__profile a {
  display: flex;
}

.sidebar__profile img {
  border-radius: 100%;
  margin: 0 auto;
  width: 100px;
}

.sidebar__profile h3 {
  color: var(--primary-color);
  text-transform: uppercase;
}

.sidebar__inner {
  background-color: var(--module-color);
  border-radius: 5px;
  height: 100%;
  width: 100%;
  position: relative;
}

.sidebar__inner ul {
  display: flex;
  flex-direction: column;
  padding: var(--space-md);
}

.sidebar__inner li {
  height: 2rem;
  display: flex;
  align-items: center;
  background-color: var(--module-color);
  color: #e3eaf8;
  border-top: 1px solid var(--white-color);
  padding: var(--space-lg);
}

.sidebar__inner li svg {
  margin: 0 10px;
}

.grid_container {
  flex: 1;
  display: flex;
  flex-direction: column;
  height: 100%;
  overflow: auto;
  padding: var(--space-md);
  margin-left: var(--dashboard-button-width);
}

@media (min-width: 768px) {
  .sidebar {
    position: relative;
    transform: none;
    width: 300px;
  }

  .toggle {
    display: none;
  }

  .sidebar {
    position: relative;
    transform: translateX(0%);
    width: 300px;
  }

  .grid_container {
    margin-left: auto;
  }
}
