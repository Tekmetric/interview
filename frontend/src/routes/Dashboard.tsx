import React from "react";
import {
  Divider,
  FormControl,
  InputLabel,
  MenuItem,
  OutlinedInput,
  Select,
  SelectChangeEvent,
  Typography,
} from "@mui/material";
import { MainContent } from "../components/Layout";
import { useCountries } from "../context/CountriesContext";
import { Country } from "../interfaces/country";
import { CountryInfo } from "../components/countryComponents/CountryInfo";
import { CountryHolidays } from "../components/countryComponents/CountryHolidays";

export const Dashboard = () => {
  const { countries } = useCountries();
  const [selectedCountry, setSelectedCountry] = React.useState<Country | null>(
    null
  );

  const handleChange = (event: SelectChangeEvent) => {
    const {
      target: { value },
    } = event;
    setSelectedCountry(countries.find((c) => c.countryCode === value) ?? null);
  };

  return (
    <MainContent>
      <Typography variant="subtitle1">
        Available Holidays information!
      </Typography>
      <FormControl sx={{ m: 1, width: 300 }}>
        <InputLabel id="select-country">Select Country</InputLabel>
        <Select
          value={selectedCountry?.name || ""}
          onChange={handleChange}
          label="Select Country"
          input={
            <OutlinedInput
              label="Select Country"
              value={selectedCountry?.name || ""}
            />
          }
          renderValue={(selected) => {
            if (selected.length === 0) {
              return <em>Select Country</em>;
            }

            return selected;
          }}
        >
          {countries.map(({ countryCode, name }) => (
            <MenuItem key={countryCode} value={countryCode}>
              {name}
            </MenuItem>
          ))}
        </Select>
      </FormControl>
      {selectedCountry && (
        <>
          <CountryInfo country={selectedCountry} />
          <Divider />
          <CountryHolidays country={selectedCountry} />
        </>
      )}
    </MainContent>
  );
};
