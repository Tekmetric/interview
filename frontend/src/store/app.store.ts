import { configureStore } from "@reduxjs/toolkit";
import sessionReducer from './session.store';
import sightingsReducer from './sightings.store';

export const store = configureStore({
  reducer: {
    session: sessionReducer,
    sightings: sightingsReducer,
  }
});

export type AppStore = typeof store;
export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
