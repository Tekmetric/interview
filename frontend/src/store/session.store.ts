import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { defaultTheme } from "../constants/theme.constants";
import { Themes } from "../types/Theme";

export interface SessionState {
  username: string | undefined,
  isAuthenticated: boolean,
  theme: Themes,
};

const initialState: SessionState = {
  username: undefined,
  isAuthenticated: false,
  theme: defaultTheme
};

export const sessionSlice = createSlice({
  name: 'session',
  initialState,
  reducers: {
    setTheme: (state, action: PayloadAction<Themes>) => {
      state.theme = action.payload;
    },
    completeLogin: (state, action: PayloadAction<string>) => {
      state.username = action.payload;
      state.isAuthenticated = !!action.payload;
    },
    logout: (state) => {
      state.username = undefined;
      state.isAuthenticated = false;
    }
  }
})

export const { setTheme, completeLogin, logout } = sessionSlice.actions;

export default sessionSlice.reducer
