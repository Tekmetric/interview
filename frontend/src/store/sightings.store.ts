import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { Sighting } from "../types/Sighting";

export const DEFAULT_ITEMS_PER_PAGE = 10;

export interface SightingState {
  sightings: Sighting[],
  currentPage: number;
  itemsPerPage: number;
  totalPages: number;
};

const initialState: SightingState = {
  sightings: [],
  currentPage: 0,
  itemsPerPage: DEFAULT_ITEMS_PER_PAGE,
  totalPages: 0,
};

export const sightingSlice = createSlice({
  name: 'sighting',
  initialState,
  reducers: {
    setSightings: (state, action: PayloadAction<Sighting[]>) => {
      state.sightings = action.payload;
    },
    setTotalPages: (state, action: PayloadAction<number>) => {
      state.totalPages = action.payload;
    },
    previousPage: (state) => {
      state.currentPage = Math.max(0, state.currentPage - 1);
    },
    nextPage: (state) => {
      state.currentPage = Math.min(state.totalPages -1, state.currentPage + 1);
    }
  }
})

export const { setSightings, setTotalPages, nextPage, previousPage } = sightingSlice.actions;

export default sightingSlice.reducer
