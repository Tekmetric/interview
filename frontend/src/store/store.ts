import { configureStore } from '@reduxjs/toolkit';
import { Reducer, UnknownAction } from 'redux';
import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import { ApplicationState } from '../features/application/types';
import { VehiclesState } from '../features/vehicles/types';
import vehicleReducer from '../features/vehicles/reducer';
import applicationReducer from '../features/application/reducer';

// Casting reducers to fix RTK type mismatch with classic reducer setup
const store = configureStore({
  reducer: {
    application: applicationReducer as Reducer<ApplicationState, UnknownAction>,
    vehicles: vehicleReducer as Reducer<VehiclesState, UnknownAction>,
  },
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export default store;

export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
