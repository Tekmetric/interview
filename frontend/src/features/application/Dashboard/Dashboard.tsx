import type { FC } from 'react';
import React, { useState } from 'react';
import { Link } from 'react-router-dom';
import { FaUser, FaCar, FaCalendarCheck } from 'react-icons/fa';

import { useAppSelector, useAppDispatch } from '../../../store/store';

import NoticeBanner from '../NoticeBanner';
import styles from './dashboard.module.css';

interface DashboardProps {
  children: React.ReactNode;
}

const Dashboard: FC<DashboardProps> = ({ children }) => {
  const user = useAppSelector((state) => state.application.user);

  const [menuOpen, setMenuOpen] = useState(false);

  return (
    <div className={styles.dashboard}>
      <NoticeBanner />

      <section className={`${styles.sidebar} ${menuOpen ? styles.active : undefined}`}>
        <button className={styles.toggle} onClick={() => setMenuOpen(!menuOpen)}>
          â˜°
        </button>
        <div className={styles.sidebar__inner}>
          <Link className={styles.logo} to={'/'}>
            <span>Tekmetric</span>
            <span>Candidate Screen</span>
          </Link>

          <div className={styles.sidebar__profile}>
            <Link to="/profile">
              <img
                src={`${user.image}`}
                alt={`user profile, ${user.name}`}
                onError={({ currentTarget }) => {
                  currentTarget.onerror = null;
                  currentTarget.src = 'https://placehold.co/300';
                }}
              />
            </Link>
            <h2>{user.name}</h2>
            <h3>{user.role}</h3>
          </div>
          <ul>
            <li>
              <FaCar aria-hidden="true" />
              <Link to="/">Inventory</Link>
            </li>
            <li>
              <FaCalendarCheck aria-hidden="true" />
              <Link to="/scheduling">Schedules</Link>
            </li>
            <li>
              <FaUser aria-hidden="true" />
              <Link to="/profile">Profile</Link>
            </li>
          </ul>
        </div>
      </section>
      <section className={styles.grid_container}>{children}</section>
    </div>
  );
};

export default Dashboard;
