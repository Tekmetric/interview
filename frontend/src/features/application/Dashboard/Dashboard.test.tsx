import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import store from '../../../store/store';
import Dashboard from './Dashboard';

describe('Dashboard', () => {
  it('renders correctly', () => {
    const { container } = render(
      <Provider store={store}>
        <MemoryRouter>
          <Dashboard>
            <p>rendered</p>
          </Dashboard>
        </MemoryRouter>
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });
});
