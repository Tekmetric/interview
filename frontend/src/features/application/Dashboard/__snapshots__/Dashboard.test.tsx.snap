// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Dashboard > renders correctly 1`] = `
<div>
  <div
    class="_dashboard_e01829"
  >
    <section
      class="_sidebar_e01829 undefined"
    >
      <button
        class="_toggle_e01829"
      >
        â˜°
      </button>
      <div
        class="_sidebar__inner_e01829"
      >
        <a
          class="_logo_e01829"
          data-discover="true"
          href="/"
        >
          <span>
            Tekmetric
          </span>
          <span>
            Candidate Screen
          </span>
        </a>
        <div
          class="_sidebar__profile_e01829"
        >
          <a
            data-discover="true"
            href="/profile"
          >
            <img
              alt="user profile, Eric Callan"
              src="https://ericcallantemp.s3.us-east-2.amazonaws.com/profiles/profile_ericcallan.jpeg"
            />
          </a>
          <h2>
            Eric Callan
          </h2>
          <h3>
            candidate
          </h3>
        </div>
        <ul>
          <li>
            <svg
              aria-hidden="true"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M499.99 176h-59.87l-16.64-41.6C406.38 91.63 365.57 64 319.5 64h-127c-46.06 0-86.88 27.63-103.99 70.4L71.87 176H12.01C4.2 176-1.53 183.34.37 190.91l6 24C7.7 220.25 12.5 224 18.01 224h20.07C24.65 235.73 16 252.78 16 272v48c0 16.12 6.16 30.67 16 41.93V416c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32v-32h256v32c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32v-54.07c9.84-11.25 16-25.8 16-41.93v-48c0-19.22-8.65-36.27-22.07-48H494c5.51 0 10.31-3.75 11.64-9.09l6-24c1.89-7.57-3.84-14.91-11.65-14.91zm-352.06-17.83c7.29-18.22 24.94-30.17 44.57-30.17h127c19.63 0 37.28 11.95 44.57 30.17L384 208H128l19.93-49.83zM96 319.8c-19.2 0-32-12.76-32-31.9S76.8 256 96 256s48 28.71 48 47.85-28.8 15.95-48 15.95zm320 0c-19.2 0-48 3.19-48-15.95S396.8 256 416 256s32 12.76 32 31.9-12.8 31.9-32 31.9z"
              />
            </svg>
            <a
              data-discover="true"
              href="/"
            >
              Inventory
            </a>
          </li>
          <li>
            <svg
              aria-hidden="true"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M436 160H12c-6.627 0-12-5.373-12-12v-36c0-26.51 21.49-48 48-48h48V12c0-6.627 5.373-12 12-12h40c6.627 0 12 5.373 12 12v52h128V12c0-6.627 5.373-12 12-12h40c6.627 0 12 5.373 12 12v52h48c26.51 0 48 21.49 48 48v36c0 6.627-5.373 12-12 12zM12 192h424c6.627 0 12 5.373 12 12v260c0 26.51-21.49 48-48 48H48c-26.51 0-48-21.49-48-48V204c0-6.627 5.373-12 12-12zm333.296 95.947l-28.169-28.398c-4.667-4.705-12.265-4.736-16.97-.068L194.12 364.665l-45.98-46.352c-4.667-4.705-12.266-4.736-16.971-.068l-28.397 28.17c-4.705 4.667-4.736 12.265-.068 16.97l82.601 83.269c4.667 4.705 12.265 4.736 16.97.068l142.953-141.805c4.705-4.667 4.736-12.265.068-16.97z"
              />
            </svg>
            <a
              data-discover="true"
              href="/scheduling"
            >
              Schedules
            </a>
          </li>
          <li>
            <svg
              aria-hidden="true"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M224 256c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z"
              />
            </svg>
            <a
              data-discover="true"
              href="/profile"
            >
              Profile
            </a>
          </li>
        </ul>
      </div>
    </section>
    <section
      class="_grid_container_e01829"
    >
      <p>
        rendered
      </p>
    </section>
  </div>
</div>
`;
